plugin/edit.py

```diff
@@ -55,6 +86,15 @@ def apply_change(self, region: sublime.Region, replacement: str, edit: sublime.E
             else:
                 self.view.erase(edit, region)
 
+    def parse_snippet(self, replacement: str) -> Optional[Tuple[str, Tuple[int, int]]]:
```

Comment:
I haven't taken a closer look to understand every part of this PR, but I wonder is there a particular reason for all this manual snippet handling? Also can it handle snippets with multiple tabstops correctly? If I understand the code above correcty, it manually just adds a selection for each tabstop. But this is not how snippets work, if there are multiple tabstops it should set a single cursor and then with <kbd>tab</kbd> you can jump to the next one.

I think instead of `view.insert(...)` the logic to apply text edits should better user the built-in commands instead (at least for snippets, to handle them correctly). Like in
https://github.com/sublimelsp/LSP/blob/36871c2a027d60e4fe77aa11ba4d26a8905bc511/plugin/completion.py#L354-L357

Reply:
It's obviously tailored for rust-analyzer and taken from it so this implementation assumes that there is only a single placeholder (specifically `$0` or `${0:...}`). So no multiple tab stops. But yes, those are not so much snippets but more a functionality to set cursor(s). The rust-analyzer-initiated [LSP protocol feature request](https://github.com/microsoft/language-server-protocol/issues/724) even calls it "Allow CodeActions to specify cursor position".

I remember pretty well that we've tried `insert_snippet` (or @rwols did) when implementing the original code and it didn't work at all since it contains a lot of extra magic that for example auto-figures indentation. We need to do a raw insert or replace that won't do any of that.

And we need to set the cursor after applying the edit. Otherwise the selection will shift randomly.

Reply:
I see. I guess it could be tried like this for now, but I just read through that thread and I saw that there was already an example in https://github.com/microsoft/language-server-protocol/issues/724#issuecomment-483206386 where they consider to use it as a "regular" snippet with multiple tab stops. So dependent on how this will end up in the specs, it's probably only a matter of time until a server will use the snippet in this way. For the autocompletion the indentation problem is solved by client announcing [insertTextMode](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#insertTextMode) capability, where this client only supports `adjustIndentation`. I guess something like this would be needed as well for the snippet text edits then. Or alternatively they should introduce a new "simplified snippet" structure that only supports a single tab marker (or guarantee this in some other way in the specs).

---

I wonder don't we need to set `{ "snippetTextEdit": boolean }` experimental client capability for this to work, which is mentioned in the docs from rust-analyzer?

Reply:
> I wonder don't we need to set `{ "snippetTextEdit": boolean }` experimental client capability for this to work, which is mentioned in the docs from rust-analyzer?

Looks like they don't explicitly check for the capability for this "move item" functionality (https://github.com/rust-lang/rust-analyzer/blob/f8eac19b3354722a6fa0177968af54a58bb5b9e1/crates/ide/src/move_item.rs#L139-L165). They do check it many other places but in that case I wouldn't go and enable it without first checking that we correctly handle it in all cases (which we probably don't).

Reply:
Okay. I also realized that the proper place would probably be in the "experimental_capabilities" client config for rust-analyzer anyway.